error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MonoZipOfMonoVoidUsage.java

```diff
@@ -0,0 +1,163 @@
+ * of the empty publisher and prematurely terminates the reactive chain from the execution. In most
+ * cases this is not the desired behaviour.
+ *
+ * @apiNote {@code Mono<?>.zipWith(Mono<Void>)} is allowed by the Reactor API, but it is an
+ *     incorrect usage of the API. It will be flagged by ErrorProne but the fix won't be supplied.
+ *     The problem with the original code should be revisited and fixed in a structural manner by
+ *     the developer.
+ */
+@AutoService(BugChecker.class)
+@BugPattern(
+    summary =
+        "`Mono#zip` and `Mono#zipWith` should not be executed against `Mono#empty` or `Mono<Void>` parameter; "
+            + "please revisit the parameters used and make sure to supply correct publishers instead",
+    link = BUG_PATTERNS_BASE_URL + "MonoZipOfMonoVoidUsage",
+    linkType = CUSTOM,
+    severity = ERROR,
+    tags = LIKELY_ERROR)
+public final class MonoZipOfMonoVoidUsage extends BugChecker
+    implements MethodInvocationTreeMatcher {
+  private static final long serialVersionUID = 1L;
+  private static final Supplier<Type> MONO_SUPPLIER = type("reactor.core.publisher.Mono");
+
+  /**
+   * In fact, we use {@code Mono<Void>} everywhere in codebases instead of {@code Mono<Object>}
+   * (actual return type of {@link Mono#empty()}) to represent empty publisher.
+   */
+  private static final Supplier<Type> MONO_VOID_TYPE_SUPPLIER =
+      VisitorState.memoize(generic(MONO_SUPPLIER, type(Void.class.getCanonicalName())));
+
+  private static final String MONO_ZIP_WITH_METHOD = "zipWith";
+  private static final String MONO_ZIP_METHOD = "zip";
+  private static final String MONO_EMPTY_METHOD = "empty";
+  private static final Matcher<ExpressionTree> ANY_MONO_VOID_IN_PUBLISHERS =
+      anyOf(
+          allOf(
+              instanceMethod().onDescendantOf(MONO_SUPPLIER).named(MONO_ZIP_WITH_METHOD),
+              toType(
+                  MethodInvocationTree.class,
+                  hasGenericArgumentOfExactType(MONO_VOID_TYPE_SUPPLIER))),
+          allOf(
+              instanceMethod().onDescendantOf(MONO_SUPPLIER).named(MONO_ZIP_WITH_METHOD),
+              toType(
+                  MethodInvocationTree.class,
+                  argument(0, staticMethod().onClass(MONO_SUPPLIER).named(MONO_EMPTY_METHOD)))),
+          onClassWithMethodName(MONO_VOID_TYPE_SUPPLIER, MONO_ZIP_WITH_METHOD),
+          allOf(
+              staticMethod().onClass(MONO_SUPPLIER).named(MONO_ZIP_METHOD),
+              toType(
+                  MethodInvocationTree.class,
+                  hasGenericArgumentOfExactType(MONO_VOID_TYPE_SUPPLIER))),
+          allOf(
+              staticMethod().onClass(MONO_SUPPLIER).named(MONO_ZIP_METHOD),
+              toType(
+                  MethodInvocationTree.class,
+                  hasArguments(
+                      AT_LEAST_ONE,
+                      staticMethod().onClass(MONO_SUPPLIER).named(MONO_EMPTY_METHOD)))));
```

Comment:
Honestly, did not look deeper if we need all of these matchers, but they definetly _can_ be simplified. I was able to re-write into something like this, WDYT ?
```suggestion
  private static final Supplier<Type> MONO_VOID_TYPE_SUPPLIER =
      VisitorState.memoize(
          generic(type(Mono.class.getCanonicalName()), type(Void.class.getCanonicalName())));

  private static final Matcher<ExpressionTree> MONO_ZIP_AND_MONO_ZIP_WITH =
      anyOf(
          instanceMethod().onDescendantOf(MONO_VOID_TYPE_SUPPLIER).named("zipWith"),
          staticMethod().onClass(MONO_VOID_TYPE_SUPPLIER).named("zip"));

  private static final Matcher<ExpressionTree> GENERIC_ARGUMENT_DERIVED_FROM_MONO_TYPE =
      toType(MethodInvocationTree.class, hasGenericArgumentOfExactType(MONO_VOID_TYPE_SUPPLIER));
  private static final Matcher<ExpressionTree> HAS_MONO_EMPTY_AS_ARGUMENT =
      toType(
          MethodInvocationTree.class,
          hasArguments(
              AT_LEAST_ONE, staticMethod().onClass(MONO_VOID_TYPE_SUPPLIER).named("empty")));
  private static final Matcher<ExpressionTree> OPERATORS_WITH_MONO_VOID_GENERIC_ARGUMENT =
      allOf(
          MONO_ZIP_AND_MONO_ZIP_WITH,
          anyOf(GENERIC_ARGUMENT_DERIVED_FROM_MONO_TYPE, HAS_MONO_EMPTY_AS_ARGUMENT));

  private static final Matcher<ExpressionTree> ANY_MONO_VOID_IN_PUBLISHERS =
      anyOf(
          OPERATORS_WITH_MONO_VOID_GENERIC_ARGUMENT,
          onClassWithMethodName(MONO_VOID_TYPE_SUPPLIER, "zipWith"));
```

Reply:
Feel free to propose better constant names if you agree with this 👍

Reply:
Yes, makes sense to me.